### PR TITLE
Add variables support to mapper

### PR DESCRIPTION
**Pull request by webrouse**
Part of: https://keboola.atlassian.net/browse/COM-1137

Changes:
- implemented `mapper.variables`.

### CODE REVIEW

File: internal/pkg/mapper/variables/load.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package variables
+
+import "github.com/keboola/keboola-as-code/internal/pkg/model"
+
+func (m *variablesMapper) MapAfterRemoteLoad(recipe *model.RemoteLoadRecipe) error {
```
Comment (webrouse) [inline internal/pkg/mapper/variables/load.go:5]: Samotna implementacia `variables` je po tom vsetkom predchadzajucom refactoringu uz velmi jednoducha:

File: internal/pkg/mapper/variables/load.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	}
+
+	// Variables ID is stored in configuration
+	variablesIdRaw, found := apiObject.Content.Get(model.VariablesIdContentKey)
+	if !found {
+		return nil
+	}
+
+	// Variables ID must be string
+	variablesId, ok := variablesIdRaw.(string)
```
Comment (webrouse) [inline internal/pkg/mapper/variables/load.go:19]: Ak mame remote config, ktory ma kluc `variablesId` typu `string`.

File: internal/pkg/mapper/variables/load.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	if !ok {
+		return nil
+	}
+
+	// Create relation
+	internalObject := recipe.InternalObject.(*model.Config)
+	internalObject.AddRelation(&model.VariablesFromRelation{
```
Comment (webrouse) [inline internal/pkg/mapper/variables/load.go:26]: Tak vytvorime relaciu, `VariablesFromRelation`.

File: internal/pkg/mapper/variables/load.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+		Source: model.ConfigKeySameBranch{
+			ComponentId: model.VariablesComponentId,
+			Id:          variablesId,
+		},
+	})
+
+	// Remove variables ID from configuration content
+	internalObject.Content.Delete(model.VariablesIdContentKey)
```
Comment (webrouse) [inline internal/pkg/mapper/variables/load.go:34]: A zmazeme kluc z obsahu konfiguracie, ... v internej reprezentacii.

`mapper.relations`, z prev PRs sa postara o to, ze sa vytvori aj druha strana relacie:
https://github.com/keboola/keboola-as-code/tree/main/internal/pkg/mapper/relations

https://github.com/keboola/keboola-as-code/blob/aa281381d39bd5b8ad808849936a0e9181e065b0/internal/pkg/mapper/relations/load.go#L37-L58

File: internal/pkg/mapper/variables/save.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+		return fmt.Errorf(
+			`unexpected state: %s has %d relations of the "%s" type, only one allowed`,
+			recipe.Manifest.Desc(),
+			len(relations),
+			relType,
+		)
+	}
+
+	// Set variables ID
+	relation := relations[0].(*model.VariablesFromRelation)
```
Comment (webrouse) [inline internal/pkg/mapper/variables/save.go:31]: Pri `remote` `save` je proces opacny.

File: internal/pkg/mapper/variables/save.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+	apiObject := recipe.ApiObject.(*model.Config)
+	apiObject.Content.Set(model.VariablesIdContentKey, relation.Source.Id)
```
Comment (webrouse) [inline internal/pkg/mapper/variables/save.go:33]: Nastavime v konfiguracii kluc.

File: internal/pkg/mapper/variables/save.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+	// Delete relation
+	apiObject.Relations.RemoveByType(relType)
```
Comment (webrouse) [inline internal/pkg/mapper/variables/save.go:36]: A zmazeme vztah, ... to by tu nemuselo byt kedze `apiObject.Relations` pri ukladani nehraju ziadnu rolu, ale nech je v tom poriadok.

File: internal/pkg/mapper/variables/load_test.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	. "github.com/keboola/keboola-as-code/internal/pkg/mapper/variables"
+	"github.com/keboola/keboola-as-code/internal/pkg/model"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils"
+)
+
+func TestVariablesMapAfterRemoteLoad(t *testing.T) {
```
Comment (webrouse) [inline internal/pkg/mapper/variables/load_test.go:13]: Test pre `remote` `load`.

File: internal/pkg/mapper/variables/save_test.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	. "github.com/keboola/keboola-as-code/internal/pkg/mapper/variables"
+	"github.com/keboola/keboola-as-code/internal/pkg/model"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils"
+)
+
+func TestVariablesMapBeforeRemoteSave(t *testing.T) {
```
Comment (webrouse) [inline internal/pkg/mapper/variables/save_test.go:13]: Test pre `remote` `save`.

File: test/pull/expected-stdout [hunk 10]
```diff
@@ -12,10 +12,10 @@ INFO	  +  R bar/extractor/keboola.ex-db-mysql/with-rows/rows/disabled
 INFO	  +  R bar/extractor/keboola.ex-db-mysql/with-rows/rows/test-view
 INFO	  +  R bar/extractor/keboola.ex-db-mysql/with-rows/rows/users
 INFO	  +  C bar/other/keboola.variables/shared-code-variables
-INFO	  +  C bar/other/keboola.variables/transformation-with-shared-code-variables
-INFO	  +  R bar/other/keboola.variables/transformation-with-shared-code-variables/values/default-values
 INFO	  +  C bar/transformation/keboola.python-transformation-v2/python-transformation
 INFO	  +  C bar/transformation/keboola.python-transformation-v2/transformation-with-shared-code
+INFO	  +  C bar/transformation/keboola.python-transformation-v2/transformation-with-shared-code/variables
```
Comment (webrouse) [inline test/pull/expected-stdout:17]: Zmena vo funkcionalnom teste `pull`.
Konfiguracia `keboola.variables` sa uz ulozi na spravne miesto, pod config co ju pouziva.

Edge cases, ak by konfiguracia neexistovala, alebo viacere konfiguracie mali rovnake `variablesId` budu v dalsich PRs.

File: test/pull/expected-stdout [hunk 11]
```diff
@@ -12,10 +12,10 @@ INFO	  +  R bar/extractor/keboola.ex-db-mysql/with-rows/rows/disabled
 INFO	  +  R bar/extractor/keboola.ex-db-mysql/with-rows/rows/test-view
 INFO	  +  R bar/extractor/keboola.ex-db-mysql/with-rows/rows/users
 INFO	  +  C bar/other/keboola.variables/shared-code-variables
```
Comment (webrouse) [inline test/pull/expected-stdout:14]: V teste su este jedny `variables` (tu ziadna zmena), ... jedna sa o `shared code`, ktory ma `variablesId`, ale nie v configu, ale v row:
https://github.com/keboola/keboola-as-code/blob/aa281381d39bd5b8ad808849936a0e9181e065b0/internal/pkg/fixtures/configs/shared-codes.json#L23-L29

To sa musi vyriesit ako samostatny pripad.

File: test/pull/out/.keboola/manifest.json [hunk 12]
```diff
@@ -87,18 +87,6 @@
       "path": "other/keboola.variables/shared-code-variables",
       "rows": []
     },
-    {
-      "branchId": %%TEST_BRANCH_BAR_ID%%,
```
Comment (webrouse) [inline test/pull/out/.keboola/manifest.json:91]: Presunute nizsie, lebo zoradene podla `path`.

File: test/pull/out/.keboola/manifest.json [hunk 13]
```diff
@@ -113,6 +101,27 @@
       "path": "transformation/keboola.python-transformation-v2/transformation-with-shared-code",
       "rows": []
     },
+    {
+      "branchId": %%TEST_BRANCH_BAR_ID%%,
+      "componentId": "keboola.variables",
+      "id": "%%TEST_BRANCH_BAR_CONFIG_TRANSFORMATION_WITH_SHARED_CODE_VARIABLES_ID%%",
+      "path": "variables",
```
Comment (webrouse) [inline test/pull/out/.keboola/manifest.json:108]: Zmenena `path`.

File: test/pull/out/.keboola/manifest.json [hunk 14]
```diff
@@ -113,6 +101,27 @@
...
+      "relations": [
+        {
+          "target": {
```
Comment (webrouse) [inline test/pull/out/.keboola/manifest.json:111]: A pribudla `Relation` `variablesFor`.

File: test/pull/out/bar/transformation/keboola.python-transformation-v2/transformation-with-shared-code/config.json [hunk 15]
```diff
@@ -4,6 +4,5 @@
   "shared_code_row_ids": [
     "%%TEST_SHARED_CODE_ROW_ID%%"
   ],
-  "variables_id": "%%TEST_TRANSFORMATION_WITH_SHARED_CODE_VARIABLES_ID%%",
```
Comment (webrouse) [inline test/pull/out/bar/transformation/keboola.python-transformation-v2/transformation-with-shared-code/config.json:7]: V configu uz teda nie je `variables_id`, a user moze cely adresar skopirovat, a nebude mat 2 konfiguracie, ktore by odkazovali na rovnake `variables`.

File: test/create-branch/in/.keboola/manifest.json [hunk 16]
```diff
@@ -4,7 +4,7 @@
     "id": %%TEST_KBC_PROJECT_ID%%,
     "apiHost": "%%TEST_KBC_STORAGE_API_HOST%%"
   },
-  "sortBy": "id",
+  "sortBy": "path",
```
Comment (webrouse) [inline test/create-branch/in/.keboola/manifest.json:7]: Zistil som, ze v niektorych testoch bolo nastavene blbe poradie, ... zaujimave, ze to zacalo nahodne padat az teraz.

**Review by JakubMatejka — APPROVED** [review]
👍
